tensorly/decomposition/tests/test_parafac2.py

```diff
@@ -0,0 +1,172 @@
+            corr *= corr_matrix[i, j]
+        
+        if corr > best_corr:
+            best_corr = corr
+    
+    return best_corr
+
+
+@pytest.mark.parametrize("normalize_factors", [True, False])
+def test_parafac2(normalize_factors):
+    rng = check_random_state(1234)
+    tol_norm_2 = 10e-2
+    rank = 6
+
+    random_parafac2_tensor = random_parafac2(
+        shapes=[(50 + rng.randint(10),60) for _ in range(40)],
+        rank=rank,
+        random_state=rng,
+    )
+    # It is difficult to correctly identify B[i, :, r] if A[i, r] is small.
+    # This is sensible, since then B[i, :, r] contributes little to the total value of X.
+    # To test the PARAFAC2 decomposition in the precence of roundoff errors, we therefore add
+    # 0.01 to the A factor matrix.
+    random_parafac2_tensor.factors[0] += 0.01
+
+    tensor = parafac2_to_tensor(random_parafac2_tensor)
+    slices = parafac2_to_slices(random_parafac2_tensor)
+
+    rec = parafac2(slices, rank, random_state=rng, normalize_factors=normalize_factors)
+    rec_tensor = parafac2_to_tensor(rec)
+
+    error = T.norm(rec_tensor - tensor, 2)
+    error /= T.norm(tensor, 2)
+    assert_(error < tol_norm_2,
+            'norm 2 of reconstruction higher than tol')
+
+    # Test factor correlation
+    A_sign = T.sign(random_parafac2_tensor.factors[0])
+    rec_A_sign = T.sign(rec.factors[0])
+    A_corr = best_correlation(A_sign*random_parafac2_tensor.factors[0], rec_A_sign*rec.factors[0])
+    assert_(T.prod(A_corr) > 0.98**rank)
+
+    C_corr = best_correlation(random_parafac2_tensor.factors[2], rec.factors[2])
+    assert_(T.prod(C_corr) > 0.98**rank)
+
+    for i, (true_proj, rec_proj) in enumerate(zip(random_parafac2_tensor.projections, rec.projections)):
+        true_Bi = T.dot(true_proj, random_parafac2_tensor.factors[1])*A_sign[i]
+        rec_Bi = T.dot(rec_proj, rec.factors[1])*rec_A_sign[i]
+        Bi_corr = best_correlation(true_Bi, rec_Bi)
+        assert_(T.prod(Bi_corr) > 0.98**rank)
+
+
+def test_parafac2_slice_and_tensor_input():
+    rank = 6
+
+    random_parafac2_tensor = random_parafac2(
+        shapes=[(50, 60) for _ in range(40)],
+        rank=rank,
+        random_state=1234
+    )
+    tensor = parafac2_to_tensor(random_parafac2_tensor)
+    slices = parafac2_to_slices(random_parafac2_tensor)
+
+    slice_rec = parafac2(slices, rank, random_state=1234, normalize_factors=False)
+    slice_rec_tensor = parafac2_to_tensor(slice_rec)
+
+    tensor_rec = parafac2(tensor, rank, random_state=1234, normalize_factors=False)
+    tensor_rec_tensor = parafac2_to_tensor(tensor_rec)
+
+    assert tl.max(tl.abs(slice_rec_tensor - tensor_rec_tensor)) < 1e-8
+
+
+def test_parafac2_normalize_factors():
+    rng = check_random_state(1234)
+    rank = 6
+
+    random_parafac2_tensor = random_parafac2(
+        shapes=[(50 + rng.randint(10),60) for _ in range(40)],
+        rank=rank,
+        random_state=rng,
+    )
+    random_parafac2_tensor.weights[0] = 100
```

Comment:
I think this should work:
```suggestion
    tl.index_update(random_parafac2_tensor.weights, tl.index[0], 100)
```